T35078 Add 'set_complete' script

JenySadadia opened the pull request:
Depends on https://github.com/kernelci/kernelci-api/pull/163

Add a client script that marks the root node as 'complete' when all child nodes are completed.

gctucker commented on config/kernelci.conf:
```diff
@@ -25,8 +25,10 @@ lab_config: shell
 [test_report]
-email_host: your.host
-email_port: your.port
+email_host: mail.collabora.com
+email_port: 587
```
Please drop these changes.  In fact I think the whole section should be removed, but that's beyond the scope of this PR.

JenySadadia replied:
Oh, it is committed by mistake. I added it for testing only.

alexandrasp commented on config/kernelci.conf:
```diff
@@ -28,6 +28,8 @@ lab_config: shell
 email_host: your.host
 email_port: your.port
 
+[trigger_completed]
```
@JenySadadia which kind of  `conf ` could we need to `trigger_completed`?

JenySadadia replied:
Atm, we don't have any configuration needed for the service. I have added it to the file for future purposes just like we have sections for `kcidb` and `notifier`.

gctucker replied:
There's at least `db_config`:
```python
class cmd_run(Command):
    help = "Trigger an event if all child nodes are completed"
    args = [Args.db_config]
```

JenySadadia replied:
Yes, I meant we don't need any additional configurations except the ones specified in `DEFAULT`.

alexandrasp reviewed (commented):
In general the code looks good! I've used the code to automatically trigger completed events and send reports, so it is working.

gctucker commented on src/trigger_completed.py:
```diff
@@ -0,0 +1,66 @@
+                node = self._db.receive_node(sub_id)
+                if node['parent'] is None:
+                    continue
+                resp = self._db.get_completed_event(node['parent'], seconds=60)
```
See my other comment, we shouldn't be doing it this way. Passing a timeout in the API call and relying on the API to sleep is going to cause some big design issues. We should just be sleeping in the client, and then send a `POST` request to update nodes when the client has determined all the child nodes have completed.

gctucker reviewed (changes requested):
There's a problem with the design here, the sleep shouldn't be done in the API.

Also the commit message isn't very descriptive of what this does, it should mention that the parent is set to "completed" state when the service has detected that all child nodes have completed.

JenySadadia commented:
> There's a problem with the design here, the sleep shouldn't be done in the API.
> 
> Also the commit message isn't very descriptive of what this does, it should mention that the parent is set to "completed" state when the service has detected that all child nodes have completed.

My concern is what about the case when root node is in `completed` state and a new child with `pending` status is added to the API? @gctucker

JenySadadia commented:
Updated the client script to change root node status to 'completed' if all child nodes are completed.
Removed the logic to wait for the child node if incomplete.

gctucker commented:
> > There's a problem with the design here, the sleep shouldn't be done in the API.
> > Also the commit message isn't very descriptive of what this does, it should mention that the parent is set to "completed" state when the service has detected that all child nodes have completed.
> 
> My concern is what about the case when root node is in `completed` state and a new child with `pending` status is added to the API? @gctucker

I think we can leave the root node in `completed` state (or `complete`) to avoid oscillations.  It basically means, the root node has reached completion with all the child nodes at some point.  The updated time can be used to diagnose things as that would show when the root node status changed to `complete`.

With this approach, we don't need to implement anything special if a new child node is added later.  If there's some issues with this then we can review the situation later.

gctucker commented on src/set_completed.py:
```diff
@@ -0,0 +1,78 @@
+
+    def run(self):
+        sub_id = self._db.subscribe_node_channel(filters={
+            'op': ('created', 'updated'),
```
If you drop this filter, then I think you'll get all the events in the same way.

JenySadadia replied:
Yes, as now we have discarded the `completed` event, we can drop this.

gctucker commented on src/set_completed.py:
```diff
@@ -0,0 +1,78 @@
+        })
+        self._logger.log_message(logging.INFO, "Listening for node events... ")
+        self._logger.log_message(logging.INFO, "Press Ctrl-C to stop.")
+        sys.stdout.flush()
```
This can be dropped as the logger takes care of it.

JenySadadia replied:
The documentation states that the `StreamHandler.flush()` should be called explicitly at times. But I failed to get any other precise information about this.
https://docs.python.org/3/library/logging.handlers.html#logging.StreamHandler.flush

gctucker commented on src/set_completed.py:
```diff
@@ -0,0 +1,78 @@
+        finally:
+            self._db.unsubscribe(sub_id)
+
+        sys.stdout.flush()
```
Ditto

gctucker reviewed (commented):
Also I really think we should rename the state `completed` to `complete` as it's more correct.  The `timeout` state isn't a very good name either, but I can't think of a better one right now.  Maybe `expired`.

JenySadadia commented:
> > > There's a problem with the design here, the sleep shouldn't be done in the API.
> > > Also the commit message isn't very descriptive of what this does, it should mention that the parent is set to "completed" state when the service has detected that all child nodes have completed.
> > 
> > 
> > My concern is what about the case when root node is in `completed` state and a new child with `pending` status is added to the API? @gctucker
> 
> I think we can leave the root node in `completed` state (or `complete`) to avoid oscillations. It basically means, the root node has reached completion with all the child nodes at some point. The updated time can be used to diagnose things as that would show when the root node status changed to `complete`.
> 
> With this approach, we don't need to implement anything special if a new child node is added later. If there's some issues with this then we can review the situation later.

Then the test reports will display pending child nodes as well.

gctucker commented:
Adding `skip` label to get the pending test report improvements merged.

gctucker commented:
> Then the test reports will display pending child nodes as well.

No, because the parent node will be set to complete when all the child nodes are complete.  If some other pending child nodes are added later then that's another issue.

JenySadadia commented:
> Also I really think we should rename the state `completed` to `complete` as it's more correct. The `timeout` state isn't a very good name either, but I can't think of a better one right now. Maybe `expired`.

How about `lapsed` ?

gctucker commented:
Please rebase

JenySadadia commented:
Used `get_nodes` in place of `get_child_nodes_from_parent`.

gctucker commented on src/set_complete.py:
```diff
@@ -0,0 +1,73 @@
+
+
+class cmd_run(Command):
+    help = "Set root node status to 'complete' when all child nodes are \
```
To keep using the same terminology everywhere:
```suggestion
    help = "Set parent node status to 'complete' when all child nodes are \
```

JenySadadia commented on src/set_complete.py:
```diff
@@ -0,0 +1,73 @@
+                        break
+                if ret:
+                    continue
+                root_node = self._db.get_root_node(node['_id'])
```
Nope, we get a complete node in event and get sibling nodes from it. Parent node will be set to complete rather than a root node. Added a few tweaks for it.

gctucker replied:
Ah yes. I think the variable names are a bit ambiguous e.g. `node` and `nodes` could be anything.

Also `get_root_node()` is basically `get_node(node['parent'])`, maybe we don't need a special method for that.

gctucker replied:
> I think the variable names are a bit ambiguous

Ah I see you've already updated that, thanks

JenySadadia commented:
Rebased.

JenySadadia commented:
As the node schema and pipeline flow has changed a bit, don't need to merge this PR anymore.
Hence, closing it.
